src/net/send_recv/msg.rs

```diff
@@ -0,0 +1,662 @@
+            return None;
+        }
+
+        // Get the next item.
+        let item = unsafe { ptr::read_unaligned(self.data.as_ptr() as *const T) };
+
+        // Move forward.
+        let data = mem::take(&mut self.data);
+        self.data = &mut data[size_of::<T>()..];
+
+        Some(item)
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let len = self.len();
+        (len, Some(len))
+    }
+
+    fn count(self) -> usize {
+        self.len()
+    }
+
+    fn last(mut self) -> Option<Self::Item> {
+        self.next_back()
+    }
+
+    fn nth(&mut self, n: usize) -> Option<Self::Item> {
+        unsafe {
+            self.drop_items(n);
+        }
+
+        self.next()
+    }
+}
+
+impl<T> core::iter::FusedIterator for AncillaryIter<'_, T> {}
+
+impl<T> ExactSizeIterator for AncillaryIter<'_, T> {
+    fn len(&self) -> usize {
+        self.data.len() / size_of::<T>()
+    }
+}
+
+#[allow(unsafe_code)]
+impl<T> DoubleEndedIterator for AncillaryIter<'_, T> {
+    fn next_back(&mut self) -> Option<Self::Item> {
+        // See if there is a next item.
+        if self.data.len() < size_of::<T>() {
+            return None;
+        }
+
+        // Get the next item.
+        let item = unsafe {
+            let ptr = self.data.as_ptr().add(self.data.len() - size_of::<T>());
+            ptr::read(ptr as *const T)
+        };
+
+        // Move forward.
+        let len = self.data.len();
+        let data = mem::take(&mut self.data);
+        self.data = &mut data[..len - size_of::<T>()];
+
+        Some(item)
+    }
+}
+
+#[allow(unsafe_code)]
+mod messages {
+    use crate::backend::c;
+    use core::marker::PhantomData;
+    use core::ptr::NonNull;
+
+    /// An iterator over the messages in an ancillary buffer.
+    pub(super) struct Messages<'buf> {
+        /// The message header we're using to iterator over the messages.
+        msghdr: c::msghdr,
+
+        /// The current pointer to the next message header to return.
+        ///
+        /// This has a lifetime of `'buf`.
+        header: Option<NonNull<c::cmsghdr>>,
+
+        /// Capture the original lifetime of the buffer.
+        _buffer: PhantomData<&'buf mut [u8]>,
+    }
+
+    impl<'buf> Messages<'buf> {
+        /// Create a new iterator over messages from a byte buffer.
+        pub(super) fn new(buf: &'buf mut [u8], len: usize) -> Self {
+            let msghdr = {
+                let mut h: c::msghdr = unsafe { core::mem::zeroed() };
+                h.msg_control = buf.as_mut_ptr() as *mut _;
+                h.msg_controllen = len as _;
+                h
+            };
+
+            // Get the first header.
+            let header = NonNull::new(unsafe { c::CMSG_FIRSTHDR(&msghdr) });
+
+            Self {
+                msghdr,
+                header,
+                _buffer: PhantomData,
+            }
+        }
+    }
+
+    impl Iterator for Messages<'_> {
+        type Item = NonNull<c::cmsghdr>;
```

Comment:
Would it be possible make this a `&mut c::cmsghdr` instead of a raw pointer? It seems like that might simplify the uses. If not, could you add a comment mentioning what the lifetime of the pointee is?

Reply:
That sounds like a better idea! Fixed to use that.